core/src/main/java/org/apache/iceberg/DefaultValueParser.java

```diff
@@ -0,0 +1,190 @@
+        JsonNode values = defaultValue.get("values");
+        List<JsonNode> keyList = Lists.newArrayList(keys.iterator());
+        List<JsonNode> valueList = Lists.newArrayList(values.iterator());
+
+        for (int i = 0; i < keyList.size(); i++) {
+          defaultMap.put(
+              parseDefaultFromJson(type.asMapType().keyType(), keyList.get(i)),
+              parseDefaultFromJson(type.asMapType().valueType(), valueList.get(i)));
+        }
+        return defaultMap;
+      case STRUCT:
+        Preconditions.checkArgument(defaultValue.isObject(),
+            "Cannot parse %s to a %s value", defaultValue, type);
+        Map<Integer, Object> defaultStruct = Maps.newHashMap();
+        for (Types.NestedField subField : type.asStructType().fields()) {
+          String fieldIdAsString = String.valueOf(subField.fieldId());
+          Object value = defaultValue.has(fieldIdAsString) ? parseDefaultFromJson(
+              subField.type(),
+              defaultValue.get(fieldIdAsString)) : null;
+          if (value != null) {
+            defaultStruct.put(subField.fieldId(), value);
+          }
+        }
+        return defaultStruct;
+      default:
+        return null;
+    }
+  }
+
+  public static Object convertJavaDefaultForSerialization(Type type, Object value) {
+    switch (type.typeId()) {
+      case DATE:
+        return DateTimeUtil.formatEpochDays((int) value);
+      case TIME:
+        return DateTimeUtil.formatTimeOfDayMicros((long) value);
+      case TIMESTAMP:
+        return DateTimeUtil.formatEpochTimeMicros((long) value, ((Types.TimestampType) type).shouldAdjustToUTC());
+      case FIXED:
+      case BINARY:
+        return BaseEncoding.base16().encode(((ByteBuffer) value).array());
+      case LIST:
+        List<Object> defaultList = (List<Object>) value;
+        List<Object> convertedList = Lists.newArrayListWithExpectedSize(defaultList.size());
+        for (Object element : defaultList) {
+          convertedList.add(convertJavaDefaultForSerialization(type.asListType().elementType(), element));
+        }
+        return convertedList;
+      case MAP:
+        Map<Object, Object> defaultMap = (Map<Object, Object>) value;
+        Map<String, List<Object>> convertedDefault = Maps.newHashMapWithExpectedSize(2);
+        List<Object> keyList = Lists.newArrayListWithExpectedSize(defaultMap.size());
+        List<Object> valueList = Lists.newArrayListWithExpectedSize(defaultMap.size());
+        for (Map.Entry<Object, Object> entry : defaultMap.entrySet()) {
+          keyList.add(convertJavaDefaultForSerialization(type.asMapType().keyType(), entry.getKey()));
+          valueList.add(convertJavaDefaultForSerialization(type.asMapType().valueType(), entry.getValue()));
+        }
+        convertedDefault.put("keys", keyList);
+        convertedDefault.put("values", valueList);
+        return convertedDefault;
+      case STRUCT:
+        Map<Integer, Object> defaultStruct = (Map<Integer, Object>) value;
```

Comment:
This should deconstruct a `StructLike`, not a map.

Reply:
Done